user_agent/configuration.py

```diff
@@ -0,0 +1,365 @@
+        external_config['FIREFOX_VERSION'])
+    Configuration.CHROME_BUILD = external_config['CHROME_BUILD']
+    Configuration.IE_VERSION = external_config['IE_VERSION']
+    Configuration.MACOSX_CHROME_BUILD_RANGE = (
+        external_config['MACOSX_CHROME_BUILD_RANGE'])
+
+
+def load_file_configuration(json_file_path):
+    with open(json_file_path, encoding='utf-8') as data_file:
+        load_configuration(data_file)
+
+
+# --- Convertation functions ---
+
+
+def date_hook(value):
+    try:
+        return datetime.strptime(value, DATE_FORMAT)
+    except ValueError:
+        return value
+
+
+def list_to_tuple(list_of_lists):
+    return tuple(tuple(list(map(date_hook, x))) for x in list_of_lists)
+
+
+# --- Validation functions ---
+
+
+def date(fmt=DATE_FORMAT):
+    return lambda v: datetime.strptime(v, fmt)
+
+
+SCHEMA = Schema({
+    Required("OS_PLATFORM"): Schema({
+        Required('win'): All([All(Coerce(str), Length(min=1))], Length(min=1)),
+        Required('mac'): All([All(Coerce(str), Length(min=1))], Length(min=1)),
+        Required('linux'): All([All(Coerce(str), Length(min=1))],
+                               Length(min=1)),
+        Required('android'): All([All(Coerce(str), Length(min=1))],
+                                 Length(min=1))
+    }),
+    Required('FIREFOX_VERSION'): All([
+        All(ExactSequence([Coerce(str), date()]),
+            Length(min=2, max=2))], Length(min=1)),
+    Required('CHROME_BUILD'): All([All([Coerce(int)],
+                                       Length(min=3, max=3))],
+                                  Length(min=1)),
+    Required('IE_VERSION'): All([ExactSequence([Coerce(int),
+                                                Coerce(str), Coerce(str)])],
+                                Length(min=1)),
+    Required('MACOSX_CHROME_BUILD_RANGE'): Schema({str: All(
+        ExactSequence([Coerce(int), Coerce(int)]),
+        Length(min=2, max=2))}, extra=ALLOW_EXTRA)
+})
+
+
+def validate_external_configuration(external_config):
+    # Validation according to schema
+    SCHEMA(external_config)
+
+    # External validation:
+
+    # Length FIREFOX_VERSION and IE_VERSION lists
+    for browser, version_length in (
+            {'FIREFOX_VERSION': 2, 'IE_VERSION': 3}.items()):
+        for version in external_config[browser]:
+            try:
+                assert len(version) == version_length
+            except AssertionError:
+                raise MultipleInvalid(errors=[
+                    SchemaError("Too much values in {}".format(browser))
+                ])
+
+    # MACOSX_CHROME_BUILD_RANGE versions should be present in OS_PLATFORM
+    mac_platforms = list(
+        map(lambda platform_string: platform_string.split('OS X ')[1],
+            external_config['OS_PLATFORM']['mac']))
+    for version_key in external_config['MACOSX_CHROME_BUILD_RANGE'].keys():
+        try:
+            assert version_key in mac_platforms
```

Comment:
Why so complex? It could be rewritten just as
```
if version_key not in mac_platforms:
    raise Something(...)
```

Reply:
Yes, you're right I've overdone this part